agent/consul/streaming_endpoint.go

```diff
@@ -26,34 +27,62 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
 		return fmt.Errorf("only ServiceHealth topic is supported")
 	}
 
-	snapshotCh := make(chan stream.Event, 32)
-	go snapshotFunc(server.Context(), snapshotCh, req.Key)
-
 	// Resolve the token and create the ACL filter.
 	rule, err := h.srv.ResolveToken(req.Token)
 	if err != nil {
 		return err
 	}
-	filt := newACLFilter(rule, h.srv.logger, h.srv.config.ACLEnforceVersion8)
+	aclFilter := newACLFilter(rule, h.srv.logger, h.srv.config.ACLEnforceVersion8)
 
-	// Wait for the events to come in and forward them to the client.
-	for event := range snapshotCh {
-		event.SetACLRules()
-		if !filt.allowEvent(event) {
-			continue
-		}
-		if err := server.Send(&event); err != nil {
-			return err
+	// Create the boolean expression filter.
+	var eval *bexpr.Evaluator
+	if req.Filter != "" {
+		eval, err = bexpr.CreateEvaluator(req.Filter, nil)
+		if err != nil {
+			return fmt.Errorf("Failed to create boolean expression evaluator: %v", err)
 		}
 	}
 
-	// Send a marker that this is the end of the snapshot.
-	endSnapshotEvent := stream.Event{
-		Topic:   req.Topic,
-		Payload: &stream.Event_EndOfSnapshot{EndOfSnapshot: true},
-	}
-	if err := server.Send(&endSnapshotEvent); err != nil {
-		return err
+	// Send an initial snapshot of the state via events if the requested index
+	// is lower than the last sent index of the topic.
+	lastSentIndex := state.LastTopicIndex(req.Topic)
+	if req.Index < lastSentIndex {
+		snapshotCh := make(chan stream.Event, 32)
+		go snapshotFunc(server.Context(), snapshotCh, req.Key)
+
+		// Wait for the events to come in and forward them to the client.
+		for event := range snapshotCh {
+			// Filter events by ACL rules.
+			event.SetACLRules()
+			if !aclFilter.allowEvent(event) {
+				continue
+			}
+
+			// Apply boolean expression filtering.
+			if eval != nil {
+				allow, err := eval.Evaluate(event.FilterObject())
+				if err != nil {
+					return err
+				}
+				if !allow {
+					continue
+				}
+			}
+
+			// Send the event.
+			if err := server.Send(&event); err != nil {
+				return err
+			}
```

Comment:
This entire inner loop is now 20 lines of code duplicated here and in the loop below - should we refactor that into a helper method or even just inline lambda rather than continue adding to the duplication?

Reply:
Yeah, I did this after opening the PR but hadn't pushed it yet.